File: projects/order-creation-queries/app/src/main/java/com/robinfood/app/usecases/getposresolution/GetPosResolutionUseCase.java
package com.robinfood.app.usecases.getposresolution;

import com.robinfood.app.usecases.gettokenbusinesscapability.IGetTokenBusinessCapabilityUseCase;
import com.robinfood.core.dtos.posresolution.DataPosResolutionRequestDTO;
import com.robinfood.core.dtos.posresolution.GetPosResolutionsDTO;
import com.robinfood.core.enums.Result;
import com.robinfood.core.models.domain.TokenModel;
import com.robinfood.repository.posresolution.IPosResolutionRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

@Component
@Slf4j
public class GetPosResolutionUseCase implements IGetPosResolutionUseCase {

    private final IGetTokenBusinessCapabilityUseCase getTokenBusinessCapabilityUseCase;
    private final IPosResolutionRepository iPosResolutionRepository;

    public GetPosResolutionUseCase(
            IGetTokenBusinessCapabilityUseCase getTokenBusinessCapabilityUseCase,
            IPosResolutionRepository iPosResolutionRepository
    ) {
        this.getTokenBusinessCapabilityUseCase = getTokenBusinessCapabilityUseCase;
        this.iPosResolutionRepository = iPosResolutionRepository;
    }

    @Override
    public Result<GetPosResolutionsDTO> invoke(DataPosResolutionRequestDTO dataRequestDTO) {

        TokenModel token = getTokenBusinessCapabilityUseCase.invoke();

        return iPosResolutionRepository.getDataPosResolution(dataRequestDTO, token.getAccessToken());

    }
}

File: projects/posv2/app/src/main/kotlin/com/robinfood/app/usecases/getdailyreportevoucher/GetPosResolutionUseCase.kt
package com.robinfood.app.usecases.getdailyreportevoucher

import com.robinfood.core.dtos.dailyreportvoucher.DailyReportVoucherRequestDTO
import com.robinfood.core.dtos.posresolutionsequence.PosResolutionSequenceDTO
import com.robinfood.core.enums.Result
import com.robinfood.repository.dailyreportvoucher.IPosResolutionSequenceRepository
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

/**
 * Implementation of IGetPosResolutionUseCase
 */
@Service
class GetPosResolutionUseCase(
    private val posResolutionSequenceRepository: IPosResolutionSequenceRepository
) : IGetPosResolutionUseCase {

    private val log: Logger = LoggerFactory.getLogger(this.javaClass)

    override suspend fun invoke(
        dailyReportVoucherRequestDTO: DailyReportVoucherRequestDTO,
        token: String
    ): PosResolutionSequenceDTO? {

        val poosResolutionSequenceResult = posResolutionSequenceRepository.getPosResolutionSequence(
            dailyReportVoucherRequestDTO,
            token
        )

        return when (poosResolutionSequenceResult) {
            is Result.Error -> {
                log.error("Pos With Id Not Found {}", dailyReportVoucherRequestDTO.posId)
                null
            }
            is Result.Success -> poosResolutionSequenceResult.data
        }

    }

}